### PR TITLE
feat: add `timestamp` to nakamoto block header

### LINKED ISSUE #4541: Nakamoto: Add timestamp to block headers
In Stacks 2.x, Stacks blocks do not have a timestamp. This is okay because they correspond directly with bitcoin blocks, so they can be thought of as occurring at the associated bitcoin block's timestamp. In Nakamoto, this is no longer the case, so we should add a timestamp field to the Stacks block. This timestamp's relative accuracy can even be checked by signers.

My proposal is to understand timestamps as:

1. A positive value greater than the block's associated burn block (i.e., the timestamp must be greater than the timestamp of the burn block whose consensus_hash = block.consensus_hash).
2. Increasing between a tenure's blocks: a given miner should never build a block B off of block A with a smaller timestamp.

### PR DESCRIPTION
**Pull request by obycode**
### Description

`(get-block-info? time block-height)` uses this new timestamp in Clarity 3, instead of using the timestamp from the burn block. The signers will validate that this timestamp is:
- Greater than the timestamp of the previous block
- Less than 15 seconds into the future (according to their local clocks)

### Applicable issues

- fixes #4541

### Checklist

- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] Required documentation changes (e.g., `docs/rpc/openapi.yaml` and `rpc-endpoints.md` for v2 endpoints, `event-dispatcher.md` for new events)
- [ ] New clarity functions have corresponding PR in `clarity-benchmarking` repo
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

**Comment by obycode** [comment]
Remaining todo item: we should be able to call `get-tenure-info?` for the current tenure, even on the first block (related to #4789). Let me know if you'd like me to add that change to this PR, or if it's okay as a separate PR.

**Comment by obycode** [comment]
Update from today's meeting:
- Change `get-tenure-info?` to take a burn block height as its parameter instead of a tenure height
- Add `tenure-height` as a new property

I'll work on these changes now.

File: clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs
Comment (kantai) [inline clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs:708]: I think you want `check_argument_count` here -- there should be exactly two arguments.

File: clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs
Comment (kantai) [inline clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs:729]: `check_argument_count`

File: clarity/src/vm/functions/database.rs
Comment (kantai) [inline clarity/src/vm/functions/database.rs:992]:
```suggestion
    let Ok(height_value) = u32::try_from(height_value) else {
        return Ok(Value::none())
    };
```

File: clarity/src/vm/functions/database.rs
Comment (kantai) [inline clarity/src/vm/functions/database.rs:1071]:
```suggestion
    let Ok(height_value) = u32::try_from(height_value) else {
        return Ok(Value::none())
    };
```

File: clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs
Comment (obycode) [reply clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs:708]: Oops, copied from `get-block-info?` which I'm now remembering needed this for backwards compatibility.

File: clarity/src/vm/types/mod.rs
Comment (kantai) [reply clarity/src/vm/types/mod.rs:839]: Ooops -- nevermind, I misread the implementation. This typing is correct.

File: stackslib/src/net/api/postblock_proposal.rs
Comment (kantai) [inline stackslib/src/net/api/postblock_proposal.rs:338]: Drop

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/mod.rs:641]: Can you add to the docstring that this is setting the `timestamp` to roughly `Instant::now()`?

I also think that this probably isn't what we want to do-- this will set the timestamp to the start time of mining the block: shouldn't this be set at roughly the end of mining the block?

**Review by kantai — COMMENTED** [review]
Just a handful of comments, mostly LGTM!

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (obycode) [reply stackslib/src/chainstate/nakamoto/mod.rs:641]: Right, this sets it to the start time. I was thinking that at least the start time is more consistent than the end time, because we can't set it to the time just before broadcasting, since the signers will have already signed, so setting it at the start is better than setting it in the middle (after block built, before signers signed).

But maybe setting it as late as possible makes more sense. My opinion on that is weakly held.

File: clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs
Comment (obycode) [reply clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs]: Fixed in 1fad8d226e8081b4204c6a6ecc44c5ff8038013b

File: clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs
Comment (obycode) [reply clarity/src/vm/analysis/type_checker/v2_1/natives/mod.rs]: Fixed in 1fad8d226e8081b4204c6a6ecc44c5ff8038013b

File: clarity/src/vm/functions/database.rs
Comment (obycode) [reply clarity/src/vm/functions/database.rs]: Fixed in 1fad8d226e8081b4204c6a6ecc44c5ff8038013b

File: clarity/src/vm/functions/database.rs
Comment (obycode) [reply clarity/src/vm/functions/database.rs]: Fixed in 1fad8d226e8081b4204c6a6ecc44c5ff8038013b

File: stackslib/src/net/api/postblock_proposal.rs
Comment (obycode) [reply stackslib/src/net/api/postblock_proposal.rs]: Fixed in 1fad8d226e8081b4204c6a6ecc44c5ff8038013b

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (kantai) [reply stackslib/src/chainstate/nakamoto/mod.rs:642]: Yeah, I don't have a strong view on this -- this seems fine to me.

**Review by kantai — DISMISSED** [review]
LGTM

File: clarity/src/vm/database/clarity_db.rs
Comment (jcnelson) [inline clarity/src/vm/database/clarity_db.rs:928]: Can you pass `id_bhh` here so you don't have to look it up again?

File: clarity/src/vm/functions/database.rs
Comment (jcnelson) [inline clarity/src/vm/functions/database.rs:1012]: `id-block-hash` is actually short for "index block hash", not "identity block hash"

File: clarity/src/vm/docs/mod.rs
Comment (jcnelson) [inline clarity/src/vm/docs/mod.rs:1822]: This should be `id-block-hash`, since that's the nomenclature we've been using so far.

File: clarity/src/vm/docs/mod.rs
Comment (jcnelson) [inline clarity/src/vm/docs/mod.rs:1825]: This should be `block-hash`, since that's the nomenclature we've been using so far.

File: clarity/src/vm/functions/database.rs
Comment (jcnelson) [inline clarity/src/vm/functions/database.rs:1042]: I'm confused by the input domain here.  Is it supposed to take:

* a Bitcoin block height?
* a sortition height (e.g. Bitcoin block height - 666050)?
* a Stacks block height?

Asking because a lot of these fields are only meaningful in the domain of either Bitcoin or sortition heights; not so much a Stacks height (e.g. what is the `miner-spend-total` for a non-tenure-change block?)

File: clarity/src/vm/functions/database.rs
Comment (jcnelson) [inline clarity/src/vm/functions/database.rs:1088]: This seems broken. Only tenure-change blocks have VRF proofs in the `nakamoto_block_headers` table.  Furthermore, the `get_vrf_seed_for_block()` function in `HeadersDB` appears broken for Nakamoto blocks, because the column isn't called `"proof"` in `nakamoto_block_headers`; it's now called `"vrf_proof"`.

File: clarity/src/vm/functions/database.rs
Comment (obycode) [reply clarity/src/vm/functions/database.rs:1042]: The input is a Stacks block height, but the info is relevant to the tenure in which that block resides.

File: clarity/src/vm/database/clarity_db.rs
Comment (obycode) [reply clarity/src/vm/database/clarity_db.rs:928]: You mean add `id_bhh` as an optional parameter? We don't already have it at the other call sites.

File: clarity/src/vm/functions/database.rs
Comment (obycode) [reply clarity/src/vm/functions/database.rs:1088]: Okay, I see. This means that it was already broken prior to this change in a call to `get-block-info?`. I will fix it and check the other properties for similar issues.

File: clarity/src/vm/database/clarity_db.rs
Comment (jcnelson) [reply clarity/src/vm/database/clarity_db.rs:928]: Yeah, an optional parameter will do.

File: clarity/src/vm/functions/database.rs
Comment (jcnelson) [reply clarity/src/vm/functions/database.rs:1012]: Let's rename it.

Comment (obycode) [reply clarity/src/vm/functions/database.rs]: Done in e496acd2e4068216881c824dc087dab83dee632c.

File: clarity/src/vm/functions/database.rs
Comment (jcnelson) [reply clarity/src/vm/functions/database.rs:1088]: Right.  Turns out, most (all?) of these fields cannot be accessed from anything other than the tenure-start block.  Specifically:

* The `HeadersDB` implementations for Stacks chain state, when handling queries for `miner-address`, `miner-spend-total`, and `miner-spend-winner`, will all look in the `payments` table.  But, this table only contains rows pertaining to `BlockFound` tenure-change blocks in Nakamoto.  If you passed in an index block hash for a non-tenure-change Nakamoto block, your query would return no rows.

* `vrf-seed` is broken in two ways, per the above -- (1) the column name is different in Nakamoto (`vrf_proof`, not `proof`), which precludes using the `get_stacks_header_column()` function in `src/clarity_vm/mod.rs`, and (2) the VRF proof (and thus the seed, which is derived from it) is only defined for Nakamoto blocks which have a `Coinbase` transaction.  If you passed an index block hash for a non-coinbase Nakamoto block, you'd get no rows.

* `block-reward` is only defined for Nakamoto blocks with a coinbase, so it's broken for the same reason the `vrf-seed` is broken.

If `get-tenure-info` is going to take a Stacks block height, then it's going to need to resolve the identified Stacks block to its tenure-start block before issuing the above queries.

File: clarity/src/vm/database/clarity_db.rs
Comment (obycode) [reply clarity/src/vm/database/clarity_db.rs]: Added in 5b1191cf3927dabbc7516507915037f052425170.

File: stacks-common/src/types/mod.rs
Comment (jcnelson) [inline stacks-common/src/types/mod.rs:157]: This function name is a bit confusing.  Can you call it something more specific, like `has_stacks_block_timestamps()` or `has_independent_block_timestamps()`?  All Stacks blocks have timestamps; in Nakamoto, they happen to be independent of the Bitcoin block.

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/mod.rs:227]: You're going to need to provide a default value if you're going to add a non-null column (see section 4 here: https://www.sqlite.org/lang_altertable.html)

File: stackslib/src/clarity_cli.rs
Comment (jcnelson) [inline stackslib/src/clarity_cli.rs:709]: What's with the magic timestamp here?

File: stackslib/src/net/api/tests/postblock_proposal.rs
Comment (jcnelson) [inline stackslib/src/net/api/tests/postblock_proposal.rs:243]: dead code

File: stackslib/src/net/api/tests/postblock_proposal.rs
Comment (jcnelson) [inline stackslib/src/net/api/tests/postblock_proposal.rs:1]: Thanks for adding test coverage!

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (obycode) [reply stackslib/src/chainstate/nakamoto/mod.rs:227]: I think this is okay since we knew that the table is empty (changed before 3.0 starts). I tested this out.

File: stackslib/src/clarity_cli.rs
Comment (obycode) [reply stackslib/src/clarity_cli.rs:709]: Just wanted to use a real-ish value like the other method did. This is the timestamp of the 2.5 activation block. 😁

File: clarity/src/vm/functions/database.rs
Comment (obycode) [reply clarity/src/vm/functions/database.rs:1088]: Is there a clean way to get the epoch here? The simplest option is to check if we got `None` and then try again with `vrf_proof` for the column name, but if there is a better way to cheaply find our epoch or burn block height here, that seems better.

Comment (obycode) [reply clarity/src/vm/functions/database.rs:1088]: > Turns out, most (all?) of these fields cannot be accessed from anything other than the tenure-start block

Okay, I didn't realize that. I'll figure out a way to get the tenure start block based on the given stacks height.

Comment (jcnelson) [reply clarity/src/vm/functions/database.rs:1088]: Having read through this PR, there's basically zero integration test coverage of `get-stacks-block-info?` and `get-tenure-info?`.  These bugs could have been caught if you did something like this:

1. Construct a Nakamoto chain with e.g. 10 blocks per tenure
2. In each Nakamoto block, deploy a smart contract with this code body:

```clarity
(print {
   "stacks-block-height": (- stacks-block-height u1),
   "miner-address": (get-tenure-info? (- stacks-block-height u1) miner-address),
   "miner-spend-total": (get-tenure-info? (- stacks-block-height u1) miner-spend-total),
   "miner-spend-winner": (get-tenure-info? (- stacks-block-height u1) miner-spend-winner),
   "vrf-seed": (get-tenure-info? (- stacks-block-height u1) vrf-seed),
   "block-reward": (get-tenure-info? (- stacks-block-height u1) block-reward),
   "tenure-time": (get-tenure-info? (- stacks-block-height u1) time),
   "burnchain-header-hash": (get-tenure-info? (- stacks-block-height u1) burnchain-header-hash)
   "block-time": (get-stacks-block-info? (- stacks-block-height u1) time),
   "header-hash": (get-stacks-block-info? (- stacks-block-height u1) header-hash),
   "id-header-hash": (get-stacks-block-info? (- stacks-block-height u1) id-header-hash)
})
```
3. Capture each `print` event from each Nakamoto block
4. Compare each `print` event's tuple to what you'd expect to see in that Nakamoto block.

You can do something similar for the Stacks 2.x blocks leading up to Nakamoto.

**Review by jcnelson — CHANGES_REQUESTED** [review]
There are some pretty bad bugs in `get-stacks-block-info?` and `get-tenure-info?` that need to be addressed and tested.

File: clarity/src/vm/functions/database.rs
Comment (obycode) [reply clarity/src/vm/functions/database.rs:1088]: Agreed. I am already in the process of writing this test.

File: stackslib/src/net/api/tests/postblock_proposal.rs
Comment (obycode) [reply stackslib/src/net/api/tests/postblock_proposal.rs]: Removed in d5dc115ebab9307b8955ee8656def6bdb6012287.

File: stacks-common/src/types/mod.rs
Comment (obycode) [reply stacks-common/src/types/mod.rs:157]: Changed in 67f0ae2ab34b398bba4d879926fd24e771b1f487

Comment (obycode) [reply stacks-common/src/types/mod.rs]: Actually, I changed this in 0811895a9bbe19ceeb7d95eef4586a22172bf6dc to be more generic `uses_nakamoto_blocks`, then I can also use it to decide which headers table to use.

**Comment by obycode** [comment]
This is ready for re-review.

**Comment by obycode** [comment]
I just noticed an issue in `get_matured_reward`. I need to get the correct epoch for `parent_id_bhh`, as it may be different from the `child_id_bhh`. Working on fixing that now.

**Comment by obycode** [comment]
> I just noticed an issue in `get_matured_reward`. I need to get the correct epoch for `parent_id_bhh`, as it may be different from the `child_id_bhh`. Working on fixing that now.

This is fixed and the "slow" test is added. This test passes locally.

**Comment by obycode** [comment]
I opened https://github.com/stacks-network/stacks-core/issues/4887 to experiment with an appropriate timeout for that test separate from this PR.

**Review by kantai — DISMISSED** [review]

**Comment by obycode** [comment]
Resolved merge conflicts (just needed to put the new integration tests one after the other). Need another re-approval @kantai 🙏.

**Review by kantai — DISMISSED** [review]

File: stacks-common/src/types/chainstate.rs
Comment (jcnelson) [inline stacks-common/src/types/chainstate.rs:250]: Can you remove this, please?  There's nothing about this type that requires it to identify a tenure-start block.  It's just a type alias (with extra steps) for `StacksBlockId`.

Comment (obycode) [reply stacks-common/src/types/chainstate.rs:250]: I added this to get some type-safety, to ensure that a tenure-change block is passed to several of the methods, or get a compile time error.

File: stackslib/src/clarity_vm/database/mod.rs
Comment (jcnelson) [inline stackslib/src/clarity_vm/database/mod.rs:593]: Just a heads-up: this is going to break with my work on MARF'ing the chainstate.  With malleablized signatures, it's possible for multiple, different `StacksBlockId`s to be returned for a given tenure's start block.  This code will need to be revised when this change is merged.  Can you open an issue for that?

File: stackslib/src/clarity_vm/database/mod.rs
Comment (jcnelson) [inline stackslib/src/clarity_vm/database/mod.rs:551]: Broadly speaking, this is just a placeholder.  When the chainstate is fully MARF'ed, the code here will instead simply query the headers MARF for the given Stacks block's first-tenure block (literally 3 lines of code).

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/nakamoto_integrations.rs:5723]: Why have the fully-qualified type name for `Value` here when you don't use it in the type parameter for `tuple0`?

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/nakamoto_integrations.rs:5771]: But presumably, you can query the _matured_ rewards too, right?  Is this tested somewhere?

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/nakamoto_integrations.rs:5982]: Correct me if I'm wrong here, but this is only producing a single Nakamoto block, right?  We need to test that `get-stacks-block-info?` and `get-tenure-info?` work in _all_ Nakamoto blocks, not just the tenure-start block (that was the shortcoming of the previous tests).

File: stackslib/src/clarity_vm/database/mod.rs
Comment (obycode) [reply stackslib/src/clarity_vm/database/mod.rs:551]: Ok cool. So then does this negate the need for the new issue you mentioned on line 593?

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:5982]: Ah, I see you're doing that below :+1:

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:5771]: Ah, I see you have a separate test for this :+1:

**Review by jcnelson — DISMISSED** [review]
Approving for now, but please open an issue to revisit the `clarity_vm` code paths once the chainstate is fully MARF'ed.  Thanks!

File: stacks-common/src/types/chainstate.rs
Comment (jcnelson) [reply stacks-common/src/types/chainstate.rs:250]: But there's no guarantee that `TenureBlockId` is a tenure-change block.  You can cast _any_ `StacksBlockId` to a `TenureBlockId`, and the code will compile.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (obycode) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:5771]: Yes, this is checked in the test that is too slow to add to the normal CI, `check_block_info_rewards`.

File: stackslib/src/clarity_vm/database/mod.rs
Comment (jcnelson) [reply stackslib/src/clarity_vm/database/mod.rs:551]: No, please open the issue so I remember to do it after this merges :pray:

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (obycode) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs]: 🤷 fixed in d2bc1276a.

File: stacks-common/src/types/chainstate.rs
Comment (obycode) [reply stacks-common/src/types/chainstate.rs:250]: Right, but you have to actually think about it to do that cast.

Comment (obycode) [reply stacks-common/src/types/chainstate.rs:250]: Whereas without this new type, you could just pass any block and not realize that it should've been a tenure-change block. I'll add some documentation on this type (that should've been there).

Comment (obycode) [reply stacks-common/src/types/chainstate.rs:253]: Docstring added in eb9b3a182.

File: stackslib/src/clarity_vm/database/mod.rs
Comment (obycode) [reply stackslib/src/clarity_vm/database/mod.rs:551]: Opened #4900

File: stackslib/src/clarity_vm/database/mod.rs
Comment (obycode) [reply stackslib/src/clarity_vm/database/mod.rs:593]: #4900 👍

**Review by kantai — DISMISSED** [review]

**Review by jcnelson — DISMISSED** [review]

**Review by jcnelson — APPROVED** [review]

**Review by kantai — APPROVED** [review]

**Pull request merged**
